src/core/experimental/array.d

```diff
@@ -0,0 +1,1480 @@
+     * array will become empty. If `rhs` refers to the current array nothing will
+     * happen.
+     *
+     * If there are no more references to the previous array, the previous
+     * array will be destroyed; this leads to a $(BIGOH n) complexity.
+     *
+     * Params:
+     *      rhs = a reference to an array
+     *
+     * Returns:
+     *      a reference to this array
+     *
+     * Complexity: $(BIGOH n).
+     */
+    auto ref opAssign()(auto ref typeof(this) rhs)
+    {
+        if (rc.isUnique && support !is null)
+        {
+            // If this was the last reference to the payload,
+            // we can safely free
+            () @trusted { pureDeallocate(support); }();
+        }
+
+        // This will update the reference count
+        rc = rhs.rc;
+
+        support = rhs.support;
+        payload = rhs.payload;
+
+        return this;
+    }
+
+    ///
+    static if (is(T == int))
+    @safe unittest
+    {
+        auto a = rcarray!int(0);
+        auto a2 = rcarray!int(1, 2);
+
+        a = a2; // this will free the old a
+        assert(a == [1, 2]);
+        a[0] = 0;
+        assert(a2 == [0, 2]);
+    }
+
+    static if (is(T == int))
+    @safe unittest
+    {
+        auto arr = rcarray!int(1, 2, 3, 4, 5, 6);
+        auto arr1 = arr[1 .. $];
+        auto arr2 = arr[3 .. $];
+        arr1 = arr2;
+        assert(arr1 == [4, 5, 6]);
+        assert(arr2 == [4, 5, 6]);
+    }
+
+    /**
+     * Append the elements of `rhs` at the end of the array.
+     *
+     * If no allocator was provided when the list was created, the
+     * $(REF, GCAllocator, std,experimental,allocator,gc_allocator) will be used.
+     *
+     * Params:
+     *      rhs = can be an element that is implicitly convertible to `T`, an
+     *            input range of such elements, or another `Array`
+     *
+     * Returns:
+     *      a reference to this array
+     *
+     * Complexity: $(BIGOH n + m), where `m` is the number of elements in `rhs`.
+     */
+    auto ref opOpAssign(string op, U)(auto ref U rhs)
+        if (op == "~" &&
+            (is (U == typeof(this))
+             || is (U : T)
+             || (is (U == V[], V) && is(V : T))
+            ))
+    {
+        return opAssign(this ~ rhs);
+    }
+
+    ///
+    static if (is(T == int))
+    @safe unittest
+    {
+        auto a = rcarray!int(0);
+        auto a2 = rcarray!int(4, 5);
+        assert(a.length == 0);
+
+        a ~= 1;
+        a ~= [2, 3];
+        assert(a == [1, 2, 3]);
+
+        // append an input range
+        a ~= a2;
+        assert(a == [1, 2, 3, 4, 5]);
+        a2[0] = 0;
+        assert(a == [1, 2, 3, 4, 5]);
+    }
+
+    ///
+    bool opEquals(U)(const U rhs) const
+    if (is(U : const typeof(this))
+        || (is(U : const V[], V) && is(typeof(T.init == V.init))))
+    {
+        if (this.length != rhs.length) return false;
+
+        for (size_t i = 0; i < this.length; i++)
```

Comment:
Is there a reason this is not a `foreach` loop.

Reply:
Nope, leftover from orginal implementation. Fixed